File: fuzz/signal.c
#include "signal.h"
#include <stdlib.h>

static int cmp_ulong(const void *a, const void *b)
{ unsigned long x = *(const unsigned long*)a, y = *(const unsigned long*)b; return (x>y)-(x<y); }

void signal_init(struct signal_set *s, int cap)
{ s->sigs = calloc(cap, sizeof(unsigned long)); s->count = 0; s->cap = cap; }

void signal_add(struct signal_set *s, long nr, long ret)
{
	unsigned long sig = ((unsigned long)(nr & 0xffff) << 32) |
			    ((unsigned long)(ret < 0 ? -ret : 0) & 0xffffffff);
	if (s->count < s->cap) s->sigs[s->count++] = sig;
}

void signal_sort_dedup(struct signal_set *s)
{
	if (s->count <= 1) return;
	qsort(s->sigs, s->count, sizeof(unsigned long), cmp_ulong);
	int w = 1;
	for (int r = 1; r < s->count; r++)
		if (s->sigs[r] != s->sigs[r-1]) s->sigs[w++] = s->sigs[r];
	s->count = w;
}

int signal_novel(struct signal_set *a, struct signal_set *b)
{
	int i = 0, j = 0, n = 0;
	while (i < a->count && j < b->count) {
		if (a->sigs[i] == b->sigs[j]) { i++; j++; }
		else if (a->sigs[i] < b->sigs[j]) { n++; i++; } else j++;
	}
	return n + (a->count - i);
}

void signal_free(struct signal_set *s)
{ free(s->sigs); s->sigs = NULL; s->count = 0; }

File: vock/src/fuzz/signal.rs
//! Fallback signal: (syscall_nr, errno) pairs (port of fuzz/signal.c).
//!
//! syzkaller analysis.go style. Each signal packs the low 16 bits of the
//! syscall number and the magnitude of a negative return (errno) into one u64.
#![allow(dead_code)]

pub const MAX_SIGNAL: usize = 8192;

pub struct SignalSet {
    pub sigs: Vec<u64>,
    pub cap: usize,
}

impl SignalSet {
    pub fn new(cap: usize) -> SignalSet {
        SignalSet {
            sigs: Vec::new(),
            cap,
        }
    }

    pub fn count(&self) -> i32 {
        self.sigs.len() as i32
    }

    pub fn add(&mut self, nr: i64, ret: i64) {
        let errno = if ret < 0 { -ret } else { 0 } as u64 & 0xffff_ffff;
        let sig = (((nr & 0xffff) as u64) << 32) | errno;
        if self.sigs.len() < self.cap {
            self.sigs.push(sig);
        }
    }

    pub fn sort_dedup(&mut self) {
        if self.sigs.len() <= 1 {
            return;
        }
        self.sigs.sort_unstable();
        self.sigs.dedup();
    }

    /// Number of signals in `self` not present in `other` (both sorted).
    pub fn novel(&self, other: &SignalSet) -> i32 {
        let (a, b) = (&self.sigs, &other.sigs);
        let (mut i, mut j, mut n) = (0usize, 0usize, 0i32);
        while i < a.len() && j < b.len() {
            if a[i] == b[j] {
                i += 1;
                j += 1;
            } else if a[i] < b[j] {
                n += 1;
                i += 1;
            } else {
                j += 1;
            }
        }
        n + (a.len() - i) as i32
    }
}
